File: Dice.py
import random

class Dice:
    def __init__(self, p_faces: list[float]):
        total = 0
        soma = 0
        self.p_faces = []
        self.accumulated_p = []
        
        for p in p_faces:
            total += p
            
        for i, p in enumerate(p_faces):
            self.p_faces.append(p/total)
            if len(self.accumulated_p) == 0:
                self.accumulated_p.append(self.p_faces[-1])
            else:
                self.accumulated_p.append(self.p_faces[-1] + self.accumulated_p[-1])
            soma += (1 + i)*p
    
        self.predicted_mean = soma/total
        
    def balanced(sides: int):
        return Dice([1 for x in range(sides)])
    
    def loaded(sides: int):
        return Dice([random.random() for x in range(sides)])
    
    def high_loaded(sides: int):
        return Dice([x+1 for x in range(sides)])
    
    def low_loaded(sides: int):
        aux = [x+1 for x in range(sides)]
        aux.reverse()
        return Dice(aux)
    
    def roll(self) -> int:
        rn = random.random()
        
        for i, p in enumerate(self.accumulated_p):
            if rn <= p:
                return i+1
        return len(self.p_faces)
    
    def roll_times(self, times: int) -> list[int]:
        result = []
        for x in range(times):
            result.append(self.roll())
        return result
        
    
    def __str__(self):
        return f"Probabilty of each face: {self.p_faces}\nPredicted mean: {self.predicted_mean}"

File: Dice.java
import java.util.Random;

public class Dice {
	private float[] probabillitiesFaces;
	public float predictedMean;
	
	public Dice(float[] probabillitiesFaces) {
		this.probabillitiesFaces = probabillitiesFaces;
		this.predictedMean = 0;
		float total = 0;
		for(float p : this.probabillitiesFaces) {
			total += p;
		}
		for(int i=0; i<this.probabillitiesFaces.length; i++) {
			this.probabillitiesFaces[i] /= total;
			this.predictedMean += probabillitiesFaces[i]*(1 + i);
		}

	}
	
	public static Dice balanced(int sides) {
		float[] pFaces = new float[sides];
		for(int i=0; i<sides; i++) {
			pFaces[i] = 1;
		}
		return new Dice(pFaces);
	}
	
	public static Dice loaded(int sides) {
		Random random = new Random();
		float[] pFaces = new float[sides];
		for(int i=0; i<sides; i++) {
			pFaces[i] = random.nextInt(1000);
		}
		return new Dice(pFaces);
	}
	
	public static Dice highLoaded(int sides) {
		float[] pFaces = new float[sides];
		for(int i=0; i<sides; i++) {
			pFaces[i] = i+1;
		}
		return new Dice(pFaces);
	}
	
	public static Dice lowLoaded(int sides) {
		float[] pFaces = new float[sides];
		for(int i=0; i<sides; i++) {
			pFaces[i] = sides-i-1;
		}
		return new Dice(pFaces);
	}
	
	public int roll() {
		Random random = new Random();
		float value = random.nextFloat();
		float pTotal = 0;
		for(int i=0; i<this.probabillitiesFaces.length; i++) {
			pTotal += this.probabillitiesFaces[i];
			if(value <= pTotal) {
				return i + 1;
			}
		}
		return this.probabillitiesFaces.length;
	}
	
	public int[] rollTimes(int times) {
		int[] rolls = new int[times];
		for(int i=0; i<times; i++) {
			rolls[i] = this.roll();
		}
		return rolls;
	}
	
	public String toString() {
		String str = "Number of faces: "+this.probabillitiesFaces.length+";\n";
		str += "Probabillity for each face: [ ";
		for(float p : this.probabillitiesFaces) {
			str += p+" ";
		}
		str += "];\n";
		str += "Predicted mean: "+this.predictedMean+";";
		return str;
	}
}
